### PR TITLE
[WIP] Debug/Fix #485

### LINKED ISSUE #485: Discontinuity in trajectory
I am building a complex movment comprised of several stages. In one of them, a connecting stage using RRT sampling planner, a discontinuty appears, and the movement is taken as correct

https://github.com/ros-planning/moveit_task_constructor/assets/95743148/e4637737-23ec-4d62-a96d-76976d26cc19

I suspect that the issue might not be with MTC itself, but rather with the sampling planner configuration.

Can you point me in which direction I should look to debug this issue?

That stage is build using this:
```cpp
// Sampling planner
sampling_planner_ = std::make_shared<moveit::task_constructor::solvers::PipelinePlanner>(node_);
sampling_planner_->setProperty("goal_joint_tolerance", 1e-5);

auto stage = std::make_unique<mtc::stages::Connect>(
            name, mtc::stages::Connect::GroupPlannerVector{{"complete_chain", sampling_planner_}});

stage->setTimeout(timeout);
stage->properties().configureInitFrom(Stage::PARENT);
```

I am also getting some logs that I suspect may be related, but don't know which config parameters I should tweak
```bash
[operation_solver-14] [ERROR 1694508450.076982062] [moveit.ros_planning.planning_pipeline]: Computed path is not valid. Invalid states at index locations: [ 51 ] out of 59. Explanations follow in command line. Contacts are published on /display_contacts (generatePlan() at ./planning_pipeline/src/planning_pipeline.cpp:329)
[operation_solver-14] [INFO 1694508450.077185331] [moveit_collision_detection_fcl.collision_common]: Found a contact between 'work_object' (type 'Object') and 'ee_hilok' (type 'Robot attached'), which constitutes a collision. Contact information is not stored. (collisionCallback() at ./collision_detection_fcl/src/collision_common.cpp:319)
[operation_solver-14] [INFO 1694508450.077203612] [moveit_collision_detection_fcl.collision_common]: Collision checking is considered complete (collision was found and 0 contacts are stored) (collisionCallback() at ./collision_detection_fcl/src/collision_common.cpp:350)
[operation_solver-14] [ERROR 1694508450.077629306] [moveit.ros_planning.planning_pipeline]: Completed listing of explanations for invalid states. (generatePlan() at ./planning_pipeline/src/planning_pipeline.cpp:357)
[operation_solver-14] [INFO 1694508450.186159185] [moveit.ompl_planning.model_based_planning_context]: Planner configuration 'complete_chain' will use planner 'geometric::RRTConnect'. Additional configuration parameters will be set when the planner is constructed. (useConfig() at ./ompl_interface/src/model_based_planning_context.cpp:406)
[operation_solver-14] [INFO 1694508458.290763315] [moveit.ompl_planning.model_based_planning_context]: Planner configuration 'complete_chain' will use planner 'geometric::RRTConnect'. Additional configuration parameters will be set when the planner is constructed. (useConfig() at ./ompl_interface/src/model_based_planning_context.cpp:406)
[operation_solver-14] [INFO 1694508461.125787735] [moveit.ompl_planning.model_based_planning_context]: Planner configuration 'complete_chain' will use planner 'geometric::RRTConnect'. Additional configuration parameters will be set when the planner is constructed. (useConfig() at ./ompl_interface/src/model_based_planning_context.cpp:406)
[operation_solver-14] [INFO 1694508462.561658228] [moveit.ompl_planning.model_based_planning_context]: Planner configuration 'complete_chain' will use planner 'geometric::RRTConnect'. Additional configuration parameters will be set when the planner is constructed. (useConfig() at ./ompl_interface/src/model_based_planning_context.cpp:406)
[operation_solver-14] [WARN 1694508467.562497303] [moveit.ompl_planning.model_based_planning_context]: Solution is approximate (logPlannerStatus() at ./ompl_interface/src/model_based_planning_context.cpp:998)
[operation_solver-14] [INFO 1694508468.242949434] [moveit.ompl_planning.model_based_planning_context]: Planner configuration 'complete_chain' will use planner 'geometric::RRTConnect'. Additional configuration parameters will be set when the planner is constructed. (useConfig() at ./ompl_interface/src/model_based_planning_context.cpp:406)
[operation_solver-14] [WARN 1694508492.011518945] [ompl]: ./src/ompl/geometric/src/PathSimplifier.cpp:711 - Solution path may slightly touch on an invalid region of the state space (log() at ./ompl_interface/src/ompl_planner_manager.cpp:68)
[operation_solver-14] [INFO 1694508492.049325049] [moveit.ompl_planning.model_based_planning_context]: Planner configuration 'complete_chain' will use planner 'geometric::RRTConnect'. Additional configuration parameters will be set when the planner is constructed. (useConfig() at ./ompl_interface/src/model_based_planning_context.cpp:406)
[operation_solver-14] [WARN 1694508497.052255456] [moveit.ompl_planning.model_based_planning_context]: Solution is approximate (logPlannerStatus() at ./ompl_interface/src/model_based_planning_context.cpp:998)
[operation_solver-14] [INFO 1694508497.378308211] [moveit.ompl_planning.model_based_planning_context]: Planner configuration 'complete_chain' will use planner 'geometric::RRTConnect'. Additional configuration parameters will be set when the planner is constructed. (useConfig() at ./ompl_interface/src/model_based_planning_context.cpp:406)
[operation_solver-14] [WARN 1694508547.380569701] [moveit.ompl_planning.model_based_planning_context]: Timed out (logPlannerStatus() at ./ompl_interface/src/model_based_planning_context.cpp:994)
[operation_solver-14] [INFO 1694508547.543847603] [moveit.ompl_planning.model_based_planning_context]: Unable to solve the planning problem (solve() at ./ompl_interface/src/model_based_planning_context.cpp:791)
```

### PR DESCRIPTION
**Pull request by rhaschke**
Attempt to fix https://github.com/ros-planning/moveit_task_constructor/issues/485#issuecomment-1760606116:
- [x] Sometimes, this assertion is violated in `SolutionCollector`:
  https://github.com/ros-planning/moveit_task_constructor/blob/564804bd7bb9e320b304b8c4880d96c6ce63b915/core/src/container.cpp#L491
- [x] Solutions are enumerated multiple times (Connector stage has superfluous `compute()` calls)

### CODE REVIEW

File: core/src/container.cpp [hunk 1]
```diff
@@ -487,7 +487,8 @@ struct SolutionCollector
 			solutions.emplace_back(std::make_pair(trace, prio));
 		} else {
 			for (SolutionBase* successor : next) {
-				assert(!successor->isFailure());  // We shouldn't have invalid solutions
+				if (successor->isFailure())
+					continue;  // skip failures
```
Comment (v4hn) [inline core/src/container.cpp:491]: I tried to get my head around why this happens in some situations I encountered in the past and failed.
Can you describe a scenario where the assert will be violated with pruning active?
I'm fine with changing it because it is a step necessary if we want to have a flag that disables pruning for a whole task (any unpruned stage with a dead end on the other side of a partial solution will trigger this assert), but I do not yet understand why it can be invalid.

File: core/test/test_serial.cpp [hunk 2]
```diff
@@ -76,3 +69,18 @@ TEST_F(ConnectConnect, FailSucc) {
 
 	EXPECT_FALSE(t.plan());
 }
+
+// https://github.com/ros-planning/moveit_task_constructor/issues/485#issuecomment-1760606116
+// TODO: Solutions are enumerated multiple times
```
Comment (v4hn) [inline core/test/test_serial.cpp:74]: Did you fix this with `oit_to_enable` further above now or is this a separate issue?

File: core/src/stage.cpp [hunk 3]
```diff
@@ -865,7 +865,6 @@ template bool ConnectingPrivate::hasPendingOpposites<Interface::BACKWARD>(const
                                                                           const InterfaceState* start) const;
 
 bool ConnectingPrivate::canCompute() const {
-	ROS_DEBUG_STREAM("canCompute " << name() << ": " << pendingPairsPrinter());
```
Comment (v4hn) [inline core/src/stage.cpp]: Can we keep this in namespace `ConnectingPairs` or similar? At least keep the commented out line in please.

**Review by v4hn — COMMENTED** [review]
This was obviously a lot of tedious debugging (which I started to repeat yesterday before noticing this PR again). Thank you for these patches.

Please see my inline comments. I would love to merge this upstream, but am not clear on some details yet.

File: core/src/container.cpp (hunk 1 above)
Comment (rhaschke) [reply core/src/container.cpp:491]: I'm not sure about the reasons anymore (it's nearly half a year ago that I handled this). 
I think we cannot be sure that only valid solutions enter the start/end interfaces. For example, Connect stages do preserve all InterfaceStates, both successes and failures. We just save failures for later inspection as well.
Thus, we need to explicitly filter them here.
I remember that this was a no-brainer when encountering it. But now I don't remember a concrete example.

File: core/src/stage.cpp (hunk 3 above)
Comment (rhaschke) [reply core/src/stage.cpp]: Sure. Done.

File: core/test/test_serial.cpp (hunk 2 above)
Comment (rhaschke) [reply core/test/test_serial.cpp:74]: Sorry, this TODO is obsolete. It was resolved by oit_to_enable as you guessed.

File: core/src/container.cpp (hunk 1 above)
Comment (v4hn) [reply core/src/container.cpp:491]: > it's nearly half a year ago that I handled this

Fair enough, my fault for letting it lie around for so long.

> For example, Connect stages do preserve all InterfaceStates, both successes and failures.

Of course, but they also *prune* the paths on failure. so unless the solution path forks, the other end should not compute at all (as the interface state is pruned). I came up with an example case below where the assert is violated with forking solution paths:
g3811

But I observed a rare stochastic assert in the following situation some days ago where `turn` yielded a solution which hit the assert when propagating back the priority to the failed attempt in `approach grasp`:
screenshot240214-173210-red
In this situation there are no obvious forking solution paths (unless ComputeIK factors through a single pose in the wrapped generator, but that should not happen?) and the failure in `approach grasp` should have pruned the only connected interface start state in `turn`, so the compute call that generated the `turn` solution should not have happened.
I tried to write a test case for this, but failed to do so in reasonable time.

So TLDR: I guess dropping the assert is correct, but also obfuscates bugs in the pruning.

**Review by v4hn — APPROVED** [review]

File: core/src/container.cpp (hunk 1 above)
Comment (rhaschke) [reply core/src/container.cpp:491]: Strange that you cannot reproduce the error in a simpler unit test.
